examples/multiprovider/downloader2/planners.rs

```diff
@@ -0,0 +1,233 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+use bao_tree::{ChunkNum, ChunkRanges};
+use iroh::NodeId;
+
+use crate::Hash;
+
+/// Trait for a download planner.
+///
+/// A download planner has the option to be stateful and keep track of plans
+/// depending on the hash, but many planners will be stateless.
+///
+/// Planners can do whatever they want with the chunk ranges. Usually, they
+/// want to deduplicate the ranges, but they could also do other things, like
+/// eliminate gaps or even extend ranges. The only thing they should not do is
+/// to add new peers to the list of options.
+pub trait DownloadPlanner: Send + std::fmt::Debug + 'static {
+    /// Make a download plan for a hash, by reducing or eliminating the overlap of chunk ranges
+    fn plan(&mut self, hash: Hash, options: &mut BTreeMap<NodeId, ChunkRanges>);
```

Comment:
why is this called `options`?

Reply:
options to download from. What goes in is the set of options to download from, once plan is done it contains the actual plan. But agree, it is not a great name!